### ISSUE TITLE
Enhancement: [method-signature-style] new option allowMethodWithVarianceAnnotation

### ISSUE THREAD

**Issue by Josh-Cena** [typescript-eslint/typescript-eslint#4990]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/method-signature-style

### Description

I don't know what the full intention of this rule is, but for me, a large reason why I use this rule is because of the unsoundness of methods being bivariant. With the new variance annotation, it's possible to make this sound again. In this case, I actually prefer methods because they are shorter and look more like... methods.

I propose to expand the options as:

```json
{
  "@typescript-eslint/method-signature-style": [
    "error",
    "property",
    { "allowMethodWithVarianceAnnotation": true }
  ]
}
````

### Fail

```typescript
interface PiggyBank<CoinType> {
    add(item: CoinType): void;
}
```

### Pass

```typescript
interface PiggyBank<in CoinType> {
    add(item: CoinType): void;
}
```

### Additional Info

[TS Playground](https://www.typescriptlang.org/play?ts=4.8.0-dev.20220516&ssl=10&ssc=5&pln=10&pc=39#code/HYQwtgpgzgDiDGEAEATCAzEBXANgFzgCdwkBvAKCSqQEtg8JDNEkAFGgcw4E8AhEYAGsAPAGEA9nQAq3GBAB8ZStRUgUKABQ0GYAFxIJ02RACU+gG6SUAbmVUAvndQR4OEIWQ4IeJAA997Fx8AiIAROji4qHytipePtwBnDz8QsJQeIR0HIoAvH7WSAD0RUjA4kiMhOKE5I7koJCwCMh0RCQUKnQMTC1sycFpdAaSwDJyip0q1Gqa2hB6I0ZyZkiWNDZOjiporu6e3n5JQalhEVExTvFIif0nIemZ2XkFxaVVNXXkQA)

**Comment by Josh-Cena** [comment]
Oops, it didn't work as I like it to😢 [Methods are still hugely unsound!](https://www.typescriptlang.org/play?ts=4.8.0-dev.20220516#code/HYQwtgpgzgDiDGEAEBLYcBO4kG8BQShqwALhBgGYLIAKKA5vQJ4BCIwA1gDxpIDCAezQAVJjAgA+XASKyQAE3kAKFGTAAufkOCjxASk0A3IfIDcMwgF8LSeRHgAbEBmQOIJJAA9N+WXMWaKmqaAEQUAgIhekgAvFLGKGY2luaybh5MmnSMrOzcUCQYaPRSMV6mSAD0lUjAAkjkGAIYAIR41kA)

**Comment by bradzacher** [comment]
Yup! The issue with methods isn't the callsite typing, it's the definition site typing. They are bivariany in their arguments at the definition, but properties are not!
